drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c

```diff
@@ -0,0 +1,452 @@
+
+		if ((otp_mem_status&0x2) == OTP_READ_VALID) {
+			return 0;
+		}
+		poll += 1;
+	}
+	LOG_INF("failed poll_otp_read_valid()\n");
+	return -ENOEXEC;
+}
+
+static int poll_otp_wrdone(void)
+{
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while ((otp_mem_status != OTP_WR_DONE) && (poll < 100)) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x1) == OTP_WR_DONE) {
+			return 0;
+		}
+		poll += 1;
+	}
+
+	LOG_INF("failed poll_otp_wrdone()\n");
+	return -ENOEXEC;
+}
+
+static int req_otp_read_mode(void)
+{
+	uint32_t val = OTP_READ_MODE;
+	int32_t err;
+
+	write_word(OTP_RWSBMODE_ADDR, val);
+	err = poll_otp_ready();
+	return err;
+}
+
+
+static int req_otp_byte_write_mode(void)
+{
+	uint32_t val = OTP_BYTE_WRITE_MODE;
+	int32_t err;
+
+	write_word(OTP_RWSBMODE_ADDR,  val);
+	err = poll_otp_ready();
+
+	if (err) {
+		return err;
+	}
+	return 0;
+
+}
+
+static uint32_t read_otp_location_disp(uint32_t offset, uint32_t *read_val)
+{
+
+	uint32_t addr = OTP_RDENABLE_ADDR;
+	uint32_t rd_addr = OTP_READREG_ADDR;
+	int32_t err;
+
+	write_word(addr,  offset);
+	err = poll_otp_read_valid();
+	if (err) {
+		LOG_INF("OTP read failed\n");
+		return err;
+	}
+	read_word(rd_addr, read_val);
+	return *read_val;
+
+}
+
+static int  write_otp_location(uint32_t otp_location_offset, uint32_t otp_data)
+{
+	int ret;
+
+	write_word(OTP_WRENABLE_ADDR,  otp_location_offset);
+
+	write_word(OTP_WRITEREG_ADDR,  otp_data);
+
+	ret = poll_otp_wrdone();
+
+	return ret;
+}
+
+
+static int otp_rd_voltage_1V8(void)
+{
+	uint32_t val = OTP_VOLTCTRL_1V8;
+	int32_t err;
+
+	err = req_otp_standby_mode();
+
+	if (err) {
+		LOG_INF("error in %s\n", __func__);
+		return -ENOEXEC;
+	}
+	write_word(OTP_VOLTCTRL_ADDR, val);
+	return 0;
+}
+
+
+int write_otp_memory(uint32_t otp_addr, uint32_t *write_val)
+{
+	int32_t		err;
+	int32_t		mask_val;
+
+	err = poll_otp_ready();
+	if (err) {
+		LOG_INF("err in otp ready poll\n");
+		return -ENOEXEC;
+	}
+
+	set_otp_timing_reg_40mhz();
+
+	err = otp_wr_voltage_2V5();
+
+	if (err) {
+		LOG_INF("error in write_voltage 2V5\n");
+		goto _exit_otp_write;
+	} else {
+		err = req_otp_byte_write_mode();
+	}
+
+	if (err) {
+		LOG_INF("error in OTP byte write mode\n");
+		goto _exit_otp_write;
+	}
+
+	switch (otp_addr) {
+
+	case REGION_PROTECT:
+		write_otp_location(REGION_PROTECT, write_val[0]);
+		write_otp_location(REGION_PROTECT+1, write_val[0]);
+		write_otp_location(REGION_PROTECT+2, write_val[0]);
+		write_otp_location(REGION_PROTECT+3, write_val[0]);
+
+		LOG_INF("Written REGION_PROTECT0 (0x%x) : 0x%04x\n",
+						(REGION_PROTECT<<2), write_val[0]);
+		LOG_INF("Written REGION_PROTECT1 (0x%x) : 0x%04x\n",
+						(REGION_PROTECT+1)<<2, write_val[0]);
+		LOG_INF("Written REGION_PROTECT2 (0x%x) : 0x%04x\n",
+						(REGION_PROTECT+2)<<2, write_val[0]);
+		LOG_INF("Written REGION_PROTECT3 (0x%x) : 0x%04x\n",
+						(REGION_PROTECT+3)<<2, write_val[0]);
+		break;
+
+	case QSPI_KEY:
+		mask_val = QSPI_KEY_FLAG_MASK;
```

Comment:
Should this be reading the pre-existing mask value and then Oring it with QSPI_KEY_FLAG_MASK?  (Generic comment for all other cases below too)

Reply:
Not necessary since we dont require s read-modify-write operation here.